packages/gatsby-plugin-fastify/src/plugins/headers.ts

```diff
@@ -0,0 +1,46 @@
+import fp from "fastify-plugin";
+
+import { SECURITY_HEADERS } from "../utils/constants";
+import { removeQueryParmsFromUrl } from "../utils/routes";
+
+import type { Headers } from "../utils/headers";
+import type { FastifyPluginAsync, FastifyReply } from "fastify";
+
+export const handleHeaders: FastifyPluginAsync<{
+  headers: Headers;
+  useDefaultSecurity: boolean;
+}> = fp(async (fastify, { headers, useDefaultSecurity }) => {
+  fastify.addHook("onSend", async function (request, reply: FastifyReply) {
+    let pageHeaders = {};
+    const useServerHeaders = ["FUNCTION", "SSR", "PROXY"].includes(reply.mode);
+    const serverHeaders = useServerHeaders ? reply.getHeaders() : {};
+
+    // apply default security headers here instead of header-builder to
+    // prevent duplication on every page/file entry in the postbuild config
+    if (useDefaultSecurity) pageHeaders = { ...SECURITY_HEADERS["/**"] };
+
+    // get the url without query params
+    const workingURL = removeQueryParmsFromUrl(request.url);
+    const urlHeaders = headers[workingURL];
+    if (urlHeaders) pageHeaders = { ...pageHeaders, ...urlHeaders };
+
+    // handle special path type routes
+    if (["CSR", "SSR", "REDIRECT", "FUNCTION", "PROXY", "404", "500"].includes(reply.mode)) {
+      const pathHeaders = headers[reply.path];
+      if (pathHeaders) pageHeaders = { ...pageHeaders, ...pathHeaders };
+    }
+
+    // add configured headers
+    reply.headers(pageHeaders);
+
+    // check pageHeaders for "undefined" values and remove those headers
```

Comment:
I agree we shouldn't be sending undefined headers...this seems like a type checking issue somewhere else and should not need to be cleaned at runtime.

Reply:
This isn't a type checking issue. 

Perhaps you can add some input on a better way to handle this but in order to support being able to remove headers I made it so if you pass the `string` value `"undefined"` to the config object (or any of the default constants) it will properly remove the header with `reply.removeHeader`.

I originally needed this functionality to match gatsby hosting as for some routes they do not return a `cache-control` header at all, whereas Fastify was defaulting to `public, max-age: 0` if we didn't explicitly set or remove it. Afterwards, I realized it could be helpful for `customHeaders` as well to "unset" headers on specific paths and documented it in the README.

Reply:
Ah...okay. hmmm. I don't love the `"undefined"` but maybe there is no other good option. what about just using js `undefined`? or `false`? or `"unset"`?

Reply:
Also, I think we can do this in the `onPostBuild`...The only exception would be default security. but that could be cleaned up to be faster probably. Thoughts?

Reply:
> Ah...okay. hmmm. I don't love the `"undefined"` but maybe there is no other good option. what about just using js `undefined`? or `false`? or `"unset"`?

IIRC, setting the object like `{ "cache-control": undefined }` with the js `undefined` makes the object not have the `cache-control` property (it is undefined) so we can't check for it and remove the headers. I went back and forth with `"undefined"` and `"unset"` and ended on using undefined because the end result is that the header is not defined (vs the default or previously matched path headers). I think unset works too, it's just a more specific/made up term for our use-case. If you want to change it, I would lean back toward unset. 

> Also, I think we can do this in the onPostBuild...The only exception would be default security. but that could be cleaned up to be faster probably. Thoughts?

`onPostBuild` doesn't actually set or remove the headers, just puts the entry into the `headers` object, we could check previously configured defaults/customs to remove them here instead of adding the unset/undefined value - but we still have to check and `removeHeader` at runtime to cover cases where it's getting set by fastify outside of our own code, such as the `cache-control` default or other headers the user may want to remove. These lines are only looking at the specific headers object for the matched path, which is only 5-10 entries, generally, so I don't think it will make a performance difference to do some in `onPostBuild` if we still have to check/remove at runtime anyway. Thoughts?

Reply:
So, I'm guessing that `@fastify/static` is adding this header, see: https://github.com/fastify/fastify-static#send-options. If we disable that then we should not have to deal with any headers set by Fastify. There are a couple of others like `Etag` and `LastModified` but I don't see any reason these would need to be overridden. If someone brings one to light, we can address the issue then. This means the only possible things we'd need to undefine would be things the user is setting. or gatsby defaults...which is kinda why we have those as selectable options. Unless you have an immediate use-case for this I think we should shelve this as a feature. Wait till there's a clear need.